src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -265,100 +265,59 @@ protected BoundExpression LowerTest(BoundDagTest test)
 
                     case BoundDagValueTest d:
                         Debug.Assert(!input.Type.IsNullableType());
-                        return makeEqualityTest(_localRewriter.MakeLiteral(d.Syntax, d.Value, input.Type), input);
+                        return MakeValueTest(d.Syntax, input, d.Value);
 
                     case BoundDagRelationalTest d:
                         Debug.Assert(!input.Type.IsNullableType());
                         Debug.Assert(input.Type.IsValueType);
-                        return makeRelationalTest(input, d.OperatorKind, _localRewriter.MakeLiteral(d.Syntax, d.Value, input.Type));
+                        return MakeRelationalTest(d.Syntax, input, d.OperatorKind, d.Value);
 
                     default:
                         throw ExceptionUtilities.UnexpectedValue(test);
                 }
+            }
+
+            protected BoundExpression MakeValueTest(SyntaxNode syntax, BoundExpression input, ConstantValue value)
```

Comment:
Shouldn't we just remove ValueTest and use RelationalTest(Equal) instead? Seems like we always need to adjust it to get to relational logic anyways.

Reply:
They behave fundamentally different in a number of ways. Relational tests always partition the input into less/greater partitions, and obey the language semantics for the corresponding relational operator.  For equality, neither is true.  For example, `double.NaN is double.NaN` is true but `double.NaN == double.NaN` is false.  We have to use different techniques for matching them in the generated code too.  So it makes more sense to keep them separate, at least to me.